tls/s2n_send.c

```diff
@@ -169,31 +171,44 @@ ssize_t s2n_sendv_with_offset_impl(struct s2n_connection *conn, const struct iov
             }
         }
     
-        POSIX_GUARD(s2n_stuffer_rewrite(&conn->out));
-
         POSIX_GUARD(s2n_post_handshake_send(conn, blocked));
-    
+
         /* Write and encrypt the record */
-        POSIX_GUARD(s2n_record_writev(conn, TLS_APPLICATION_DATA, bufs, count, 
-            conn->current_user_data_consumed + offs, to_write));
-        conn->current_user_data_consumed += to_write;
-        conn->active_application_bytes_consumed += to_write;
-
-        /* Send it */
-        if (s2n_flush(conn, blocked) < 0) {
-            if (s2n_errno == S2N_ERR_IO_BLOCKED && user_data_sent > 0) {
-                /* We successfully sent >0 user bytes on the wire, but not the full requested payload
-                 * because we became blocked on I/O. Acknowledge the data sent. */
-
-                conn->current_user_data_consumed -= user_data_sent;
-                return user_data_sent;
-            } else {
-                S2N_ERROR_PRESERVE_ERRNO();
+        int written_to_record = s2n_record_writev(conn, TLS_APPLICATION_DATA, bufs, count, 
+            conn->current_user_data_consumed + offs, to_write);
+
+        /* Sanity check. s2n_record_writev returns how many bytes it was able to process. This should always be 
+         * a positive integer so if it less than 0 something went wrong. */
+        POSIX_ENSURE(written_to_record >= 0, S2N_ERR_ENCRYPT);
+
+        conn->current_user_data_consumed += written_to_record;
+        conn->active_application_bytes_consumed += written_to_record;
+
+        uint32_t remaining_space = s2n_stuffer_space_remaining(&conn->out);
+        
+        /* If the send is not buffered then flush every time we create a record.
+         *
+         * Otherwise flush the out buffer depending on how much data we have buffered.
+         *
+         * We assume the next record is the same size as the previous one. If the send buffer does not
+         * have enough space to store a record of the previous size, flush the send stuffer. */
+        if (((conn->send_buffer_size == 0) ||
+            ((total_size - conn->active_application_bytes_consumed == 0) || (remaining_space < written_to_record)))) {
```

Comment:
This is a very complicated conditional. You should try to simplify it and/or move it to a separate, named method.

But maybe this isn't the right place for the flush? The complicating factor here is dynamic record sizing. Why not do the flush immediately after calculating to_send instead of after actually sending the record?

But I also don't think that this logic is right. You're checking that the send buffer can fit the payload, but what about the overhead from the record + encryption? You'll probably want a send test that verifies you can completely fill your send buffer as expected.